junit/src/main/java/cucumber/runtime/junit/IdProvider.java

```diff
@@ -0,0 +1,56 @@
+package cucumber.runtime.junit;
+
+import java.io.Serializable;
+
+/**
+ * JUnit Descriptions require an unique id. However this id should be identical between test
+ * re-runs. This allows descriptions to be used to determine which failed tests should be rerun.
+ *
+ * IdProvider generates predictable unique ids. The feature name is used as a base to disambiguate
+ * between different features.
+ */
+class IdProvider {
+
+    private long id = 0;
+    private final String base;
+
+    IdProvider(String base) {
+        this.base = base;
+    }
+
+    Serializable next() {
+        return new Id(base, id++);
```

Comment:
After reviewing the changes in Gherkin v4 I've changed my opinion. It is indeed much better to use a line based notation.

 - FeatureRunner: feature-uri
 - ExecutionRunner: feature-uri:scenario-line
 - Step: feature-uri-scenario-line:step-line